### ISSUE TITLE
Reverse the order of slowlog's source and X-Opaque-Id attributes 

### ISSUE THREAD

**Issue by dblia** [elastic/elasticsearch#71557]
Hello!

Currently, in search slow log, the _X-Opaque-Id_ header is [appended as the last element of the _message_ string ](https://github.com/elastic/elasticsearch/blob/v7.12.0/server/src/main/java/org/elasticsearch/index/SearchSlowLog.java#L178) right **after** the query's _source_ attribute. Given the fact that in the _ESJsonLayout_ layout there's no control over the [message size](https://github.com/elastic/elasticsearch/blob/v7.12.0/server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java#L85), there's no guaranty that the _X-Opaque-Id_ header will exist in the slow log's output for applications that generate big queries in size. Moreover, since there's no way currently to disable the [search slow log's source](https://github.com/elastic/elasticsearch/issues/22683) similar to the [index slow log,](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-slowlog.html#index-slow-log) if you want to parse the _message_ (let's say via a grok pattern) in order to visualize the queries via a Kibana index, you can't achieve the desired behavior due to the fact that the _id_ field may have been trimmed completely from the slow log making the _X-Opaque-Id_ header a noop for those queries.

I suppose that simply reverting the order in which those entries are appended to the search slow slog could easily fix such issues. Furthermore, ensuring that the biggest slow log entry (aka _source_) is always the last element in the _message_ string would protect us from similar inconsistencies in the future, in case a probable field addition will be made in the slow log message.

---

Note that while looking for existing/similar issues I've stumbled upon this PR (#52507) which is idle (and closed) for almost a year now. Since I couldn't figure out the reason it was abandoned, I chose to open a feature request for that, since that would indeed be a really valuable modification on the search slow log functionality.

**Comment by pgomulka** [comment]
hey @dblia thank you for raising this
The user who opened #52507 decided to close it. Bear in mind that this PR was only touching the plaintext slow log files. 
Plaintext slow log files will be removed in next major version and we will only support JSON slow log logs.

The code fragment you highlighted about X-Opaque-ID in 7.12 ([appended as the last element of the message string ](https://github.com/elastic/elasticsearch/blob/v7.12.0/server/src/main/java/org/elasticsearch/index/SearchSlowLog.java#L178 )) is adding an ID to a hashmap. The order of the fields is then defined in a config file.
https://github.com/elastic/elasticsearch/blob/v7.12.0/distribution/src/config/log4j2.properties#L109
Can you try changing the order in your config and let me know if this helps?

Bear in mind in next version this might change as there will be no way to define your own order of fields. Also the JSON will be in ECS layout.

the https://github.com/elastic/elasticsearch/issues/22683 is still open for the discussion.

**Comment by pgomulka** [comment]
also @dblia can you please clarify why the `id` (x-opaque-id`) will be truncated?

**Comment by dblia** [comment]
Hello @pgomulka.

> hey @dblia thank you for raising this
> The user who opened #52507 decided to close it. Bear in mind that this PR was only touching the plaintext slow log files.
> Plaintext slow log files will be removed in next major version and we will only support JSON slow log logs.
> 
You're right about the referenced PR. However I'm actually interested in JSON-formatted slow logs so this is fine!

> The code fragment you highlighted about X-Opaque-ID in 7.12 ([appended as the last element of the message string ](https://github.com/elastic/elasticsearch/blob/v7.12.0/server/src/main/java/org/elasticsearch/index/SearchSlowLog.java#L178)) is adding an ID to a hashmap. The order of the fields is then defined in a config file.
> https://github.com/elastic/elasticsearch/blob/v7.12.0/distribution/src/config/log4j2.properties#L109
> Can you try changing the order in your config and let me know if this helps?
> 
I've mainly quoted the _v7.12_ version since it's currently the latest release. My clusters are actually on _v7.7_ and _v7.10.2_ versions so I'm afraid that I've completely missed that feature, so thank you. It seems that it covers what I describe in this feature request, so I could verify it on a _v7.12_ cluster just for completeness.

> Bear in mind in next version this might change as there will be no way to define your own order of fields. Also the JSON will be in ECS layout.
> 
I'll have to take a look at this, thanks.

> the #22683 is still open for the discussion.

:+1:

> also @dblia can you please clarify why the id (x-opaque-id`) will be truncated?
>
IIUC the reason is due to thhe _ESJsonLayout_'s maximum messages size which is currently capped to [10kb](https://github.com/elastic/elasticsearch/blob/v7.12.0/server/src/main/java/org/elasticsearch/common/logging/ESJsonLayout.java#L85). We have plenty of slow log queries that surpass that limit, subsequently the _id_ (_X-Opaque-Id_) field which follows the _source_ field in the slow log's _message_, is not included in the slow log output at all. However, [affecting the order of the message fields](https://github.com/elastic/elasticsearch/blob/v7.12.0/distribution/src/config/log4j2.properties#L109) as you suggested, should definitely solve that problem. 

So summing up, given the [_esmessagefields_](https://github.com/elastic/elasticsearch/blob/v7.12.0/distribution/src/config/log4j2.properties#L109) feature, the open discussion you referenced (#22683), and the slow logs changes we expect in the next major release (ECS layout), I'd suppose that this request could close; what do you believe?

**Comment by dblia** [comment]
Hello @pgomulka. The [esmessagefields](https://github.com/elastic/elasticsearch/blob/v7.12.0/distribution/src/config/log4j2.properties#L109) feature does the trick, as expected, so feel free to close this request. Thanks for your help.

**Comment by pgomulka** [comment]
sounds great @dblia! thanks
